browser/net/brave_proxying_url_loader_factory.h

```diff
@@ -0,0 +1,294 @@
+    void ContinueToSendHeaders(const std::set<std::string>& removed_headers,
+                               const std::set<std::string>& set_headers,
+                               int error_code);
+    void ContinueToStartRequest(int error_code);
+    void ContinueToHandleOverrideHeaders(int error_code);
+    void ContinueToResponseStarted(int error_code);
+    void ContinueToBeforeRedirect(const net::RedirectInfo& redirect_info,
+                                  int error_code);
+    void HandleResponseOrRedirectHeaders(
+        net::CompletionOnceCallback continuation);
+    void OnRequestError(const network::URLLoaderCompletionStatus& status);
+    bool IsRedirectSafe(const GURL& from_url, const GURL& to_url);
+    void HandleBeforeRequestRedirect();
+
+    // TODO(iefremov): Get rid of shared_ptr, we should clearly own the pointer.
+    std::shared_ptr<brave::BraveRequestInfo> ctx_;
+    BraveProxyingURLLoaderFactory* const factory_;
+    network::ResourceRequest request_;
+    const base::Optional<url::Origin> original_initiator_;
+    const uint64_t request_id_;
+    const int32_t network_service_request_id_;
+
+    const int render_process_id_;
+    const int frame_tree_node_id_;
+    const int32_t routing_id_;
+    const uint32_t options_;
+
+    content::ResourceContext* resource_context_;
+    const net::MutableNetworkTrafficAnnotationTag traffic_annotation_;
+    mojo::Binding<network::mojom::URLLoader> proxied_loader_binding_;
+    network::mojom::URLLoaderClientPtr target_client_;
+
+    mojo::Binding<network::mojom::URLLoaderClient> proxied_client_binding_;
+    network::mojom::URLLoaderPtr target_loader_;
+
+    // NOTE: This is state which ExtensionWebRequestEventRouter needs to have
+    // persisted across some phases of this request -- namely between
+    // |OnHeadersReceived()| and request completion or restart. Pointers to
+    // these fields are stored in a |BlockedRequest| (created and owned by
+    // ExtensionWebRequestEventRouter) through much of the request's lifetime.
+    // That code supports both Network Service and non-Network Service behavior,
+    // which is why this weirdness exists here.
+    network::ResourceResponseHead current_response_;
+    scoped_refptr<net::HttpResponseHeaders> override_headers_;
+    GURL redirect_url_;
+
+    // TODO(https://crbug.com/882661): Remove this once the bug is fixed.
+    bool on_receive_response_received_ = false;
+    bool on_receive_response_sent_ = false;
+
+    bool request_completed_ = false;
+
+    // If |has_any_extra_headers_listeners_| is set to true, the request will be
+    // sent with the network::mojom::kURLLoadOptionUseHeaderClient option, and
+    // we expect events to come through the
+    // network::mojom::TrustedURLLoaderHeaderClient binding on the factory.
+    // At the moment we always set it to true when we get a request with
+    // non-zero id. In the future we can optimize by checking if there are any
```

Comment:
TBH, I have no understanding of whether we should bother about such requests or not :-/